// node-milight/Milight.js
'use strict';

var dgram = require('dgram');
const Bridge = require('./Bridge');

const BROADCAST_IP = '255.255.255.255';
const PORT = 48899;

class Milight {

	/**
	 * Construct Milight object
	 */
	constructor() {

		// Array holding bridges found
		this.bridges = [];
	}

	/**
	 * Search for bridges and
	 * resolve array with found
	 * bridges
	 * @returns {Promise} Resolves bridges array
	 */
	getBridges() {
		return this._discoverBridges();
	}

	/**
	 * Search for specific bridge.
	 * @param id
	 * @returns {*}
	 */
	getBridge(id) {
		for (var i in this.bridges) {
			if (this.bridges[i].id === id) return this.bridges[i];
		}
	}

	/**
	 * Discover bridges
	 * @returns {Promise} Resolves with bridges array
	 * @private
	 */
	_discoverBridges() {
		return new Promise(resolve => {

			console.log("Milight: starting bridge discovery...");

			try {
				// Create dgram socket
				const socket = dgram.createSocket('udp4');

				// Bind and enable broadcast
				socket.bind(function () {
					socket.setBroadcast(true);
				});

				// When socket is listening, send discovery packets
				socket.on('listening', () => {

					// Discovery message
					const msg = new Buffer([
						0x4C, 0x69, 0x6E, 0x6B,
						0x5F, 0x57, 0x69, 0x2D,
						0x46, 0x69
					]);

					// Send discovery packets
					socket.send(msg, 0, msg.length, PORT, BROADCAST_IP);

					// Send again
					setTimeout(() => {

						// Send discovery packets
						socket.send(msg, 0, msg.length, PORT, BROADCAST_IP);
					}, 1500);

					// After 3 seconds resolve with results
					setTimeout(() => {
						resolve(this.bridges);
					}, 3000);
				});

				socket.on('message', message => {

					// Parse data from message
					const data = message.toString('ascii').split(/,|:/);
					const ip = data[0];
					const mac = data[1].replace(/(.{2})/g, "$1:").slice(0, -1);

					// Check if valid data is provided
					if (ip && mac) {

						console.log("Milight: bridge found, ip: " + ip, " mac: " + mac);

						// Store bridge internally
						this._addBridge(new Bridge({
							ip: ip,
							mac: mac,
							id: new Buffer(mac).toString('base64')
						}));
					}
				});
			} catch (err) {
				console.log("Milight: error starting discovery socket " + err);
				resolve([]);
			}
		});
	}

	pairBridge(ip) {
		console.log("Milight: starting bridge pairing on " + ip + "...");
		return new Promise((resolve, reject) => {

			try {
				// Create dgram socket
				const socket = dgram.createSocket('udp4');

				// Bind and enable broadcast
				socket.bind(function () {
					socket.setBroadcast(true);
				});

				// When socket is listening, send discovery packets
				socket.on('listening', () => {

					// Discovery message
					const msg = new Buffer([
						0x4C, 0x69, 0x6E, 0x6B,
						0x5F, 0x57, 0x69, 0x2D,
						0x46, 0x69
					]);

					// Send discovery packets
					socket.send(msg, 0, msg.length, PORT, ip);

					// After 3 seconds resolve with results
					setTimeout(() => {
						reject();
					}, 3000);
				});

				socket.on('message', message => {

					// Parse data from message
					const data = message.toString('ascii').split(/,|:/);
					const ip = data[0];
					const mac = data[1].replace(/(.{2})/g, "$1:").slice(0, -1);

					// Check if valid data is provide
					if (ip && mac) {

						console.log("Milight: bridge found, ip: " + ip, " mac: " + mac);

						// Create bridge object
						var bridge = new Bridge({
							ip: ip,
							mac: mac,
							id: new Buffer(mac).toString('base64')
						});

						// Store bridge internally
						this._addBridge(bridge);

						// Return succesful pairing
						return resolve(bridge);
					}
				});
			} catch (err) {
				console.log("Milight: bridge pairing failed " + err);
				return reject(err);
			}
		});
	}

	/**
	 * Send ping message to bridge
	 * to see if it is still online.
	 */
	pingBridge(ip) {

		// If ping timeout is already present clear it
		clearTimeout(this.pingTimeout);
		this.pingTimeout = null;

		// Set new ping timeout
		this.pingTimeout = setTimeout(()=> {

			// After two seconds consider device unavailable
			var timeout = setTimeout(()=> {

				console.log("Milight: bridge is offline");

				// Clear ping
				clearTimeout(this.pingTimeout);
				this.pingTimeout = null;

				// Keep retrying
				setTimeout(() => {
					this.pingBridge(ip);
				}, 2000);

			}, 2000);

			// Ping bridge
			tcpp.ping({address: ip}, function (err, data) {

				// Check if bridge is alive
				if (data.max && data.min) {
					clearTimeout(timeout);

					console.log("Milight: bridge: is  online");
				}

				// Clear ping
				clearTimeout(this.pingTimeout);
				this.pingTimeout = null;

			}.bind(this));
		}, 1000);
	}

	/**
	 * Adds or replaces an bridge object.
	 * @param bridge
	 * @private
	 */
	_addBridge(bridge) {
		let done = false;
		for (let i in this.bridges) {
			if (this.bridges[i].id === bridge.id) {

				// Replace it
				this.bridges[i] = bridge;
				done = true;
			}
		}

		// Add it
		if (!done) this.bridges.push(bridge);
	}
}

// Export as singleton
module.exports = new Milight();
